src/Raven.Server/Web/System/AdminGcDebugHandler.cs

```diff
@@ -0,0 +1,424 @@
+                {
+                    if (first == false)
+                        writer.WriteComma();
+
+                    first = false;
+
+                    context.Write(writer, @event.ToJson());
+                }
+
+                writer.WriteEndArray();
+                writer.WriteEndObject();
+            }
+        }
+
+        internal class GcAllocationsEventListener : Expensive_GcEventListener
+        {
+            internal const string AllocationEventName = "GCAllocationTick_V4";
+
+            private readonly Dictionary<string, AllocationInfo> _allocations = new();
+
+            public IReadOnlyCollection<AllocationInfo> Allocations => _allocations.Values;
+
+            public class AllocationInfo
+            {
+                private ulong? _allocations;
+
+                public string Type;
+                public ulong SmallObjectAllocations;
+                public ulong LargeObjectAllocations;
+                public long NumberOfSmallObjectAllocations;
+                public long NumberOfLargeObjectAllocations;
+
+                public ulong Allocations
+                {
+                    get
+                    {
+                        // used for ordering
+                        _allocations ??= SmallObjectAllocations + LargeObjectAllocations;
+                        return _allocations.Value;
+                    }
+                }
+
+                public long NumberOfAllocations => NumberOfSmallObjectAllocations + NumberOfLargeObjectAllocations;
+            }
+
+            protected override void OnEventWritten(EventWrittenEventArgs eventData)
+            {
+                switch (eventData.EventName)
+                {
+                    case AllocationEventName:
+                        var type = (string)eventData.Payload[5];
+                        if (_allocations.TryGetValue(type, out var info) == false)
+                        {
+                            _allocations[type] = info = new AllocationInfo
+                            {
+                                Type = type
+                            };
+                        }
+                        var allocations = (ulong)eventData.Payload[3];
+
+                        var smallObjectAllocation = (uint)eventData.Payload[1] == 0x0;
+                        if (smallObjectAllocation)
+                        {
+                            _allocations[type].SmallObjectAllocations += allocations;
+                            _allocations[type].NumberOfSmallObjectAllocations++;
+                        }
+                        else
+                        {
+                            _allocations[type].LargeObjectAllocations += allocations;
+                            _allocations[type].NumberOfLargeObjectAllocations++;
+                        }
+
+                        break;
+                }
+            }
+        }
+
+        private class GcEventsEventListener : Expensive_GcEventListener
```

Comment:
this was copied from somewhere? if yes, can we add a link with commit has here?

Reply:
done